src/main/scala/io/iohk/ethereum/consensus/ethash/EthashConsensus.scala

```diff
@@ -45,29 +44,31 @@ class EthashConsensus private (
     blockchainConfig = blockchainConfig
   )
 
-  private[this] val atomicMiner = new AtomicReference[Option[ActorRef]](None)
+  @volatile private[this] var minerRef: Option[ActorRef] = None
 
   private implicit val timeout: Timeout = 5.seconds
 
   override def sendMiner(msg: MinerProtocol): Future[MinerResponse] = {
     import akka.pattern.ask
-    atomicMiner
-      .get()
+    minerRef
       .map(_.ask(msg).mapTo[MinerResponse])
       .getOrElse(Future.successful(MinerNotExist))
   }
 
+  private[this] val mutex = new Object
```

Comment:
locks/mutex/semaphores are low level, declarative programming approach to concurrencly. They are hard to maintain and super hard to write correctly.

Would it be possible to use sth like STM here. Maybe Ref from cats effect: https://typelevel.org/cats-effect/concurrency/ref.html